### PR TITLE
Add equalConstantTime function for timing-safe hash comparison

### PR DESCRIPTION
**Pull request by NadhiyaNashath**
## Purpose
Add `equalConstantTime` function to compare hash values in constant time, preventing timing attacks during hash comparison. Issue - https://github.com/ballerina-platform/ballerina-library/issues/8733

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec
- [ ] Checked native-image compatibility

## Summary

This pull request adds a timing-safe comparison API for hash values and related documentation, tests, and version updates.

Core changes:
- Add public type `HashValue = byte[] | string`.
- Add public isolated function `equalConstantTime(HashValue value, HashValue expectedValue)` that compares byte arrays/strings in a timing-consistent manner.
- Add native binding `equalByteConstantTime(byte[] value, byte[] expectedValue)` mapped to Java.
- Implement Java native utility `io.ballerina.stdlib.crypto.nativeimpl.Utils.equalByteConstantTime` using Java's MessageDigest.isEqual.

Supporting updates:
- Bumped module version to 2.11.0 across build/config files (ballerina/Ballerina.toml, ballerin a/Dependencies.toml, gradle.properties).
- Updated docs/spec (docs/spec/spec.md) to add a "Utilities" section with a constant-time comparison example and adjusted numbering.
- Updated changelog to record the new API.
- Added comprehensive tests (ballerina/tests/utils_test.bal) covering functional correctness and timing-based verification of the constant-time behavior.

Tests & notes:
- Functional tests cover matching, mismatched, empty, length-mismatch, mixed-type, and HMAC integration scenarios.
- Timing tests report consistent average execution times across mismatch positions for large byte arrays and strings; additional edge-case timing and size scenarios included.
- Native-image compatibility has not been checked.

Intent:
- Provide a stable, timing-consistent comparison routine for hash values to reduce timing-related information exposure and support safer cryptographic comparisons.

## Timing Comparision Report 

2,000 iterations per test scenario.

> **Note on JVM Noise:** Fractional differences could be the result of background JVM noise rather than algorithmic short-circuiting.

#### Byte Array (100,000 bytes, 2000 iterations)
*Testing a 100kB array to verify that traversing the entire array takes a uniform amount of time regardless of where a mismatch occurs.*

| Case | Description | Avg time (µs/call) |
|------|-------------|-------------------|
| 1 | Differ at first byte | 130.67 |
| 2 | Differ at last byte | 130.57 |
| 3 | Differ at midpoint | 129.13 |
| 4 | Match (no difference) | 130.62 |

#### String (100,000 chars, 2000 iterations)
*Verifying constant-time behavior is maintained when handling string inputs.*

| Case | Description | Avg time (µs/call) |
|------|-------------|-------------------|
| 5 | Differ at first char | 144.26 |
| 6 | Differ at last char | 142.29 |
| 7 | Match (no difference) | 143.99 |

### Additional Edge Cases Tested
The implementation also passed constant-time consistency checks for the following scenarios:
* **Small payloads:** 64-byte arrays and standard HMAC-SHA256 signature lengths.
* **Massive payloads:** Arrays up to 500,000 bytes.
* **Extreme content variation:** All-zeros vs. all-255s.
* **Randomized data:** Match vs. randomly generated mismatches.

### CODE REVIEW

File: ballerina/tests/compare_test.bal [hunk 1]
```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2026 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
```
Comment (MohamedSabthar) [inline ballerina/tests/compare_test.bal]:
```suggestion
// Copyright (c) 2026 WSO2 LLC. (http://www.wso2.com).
//
// WSO2 LLC. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
```
Let's fix other places

File: ballerina/tests/utils_test.bal [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+@test:Config {}
+isolated function testEqualConstantTimeWithHmacHexStrings() {
+    byte[] input = "Hello Ballerina".toBytes();
+    byte[] key = "secret".toBytes();
+    byte[]|error hmac = hmacSha256(input, key);
+    if hmac is byte[] {
+        string value = string `sha256=${hmac.toBase16()}`;
+        string expectedValue = string `sha256=${hmac.toBase16()}`;
+        test:assertTrue(equalConstantTime(value, expectedValue));
+    }
```
Comment (gayaldassanayake) [inline ballerina/tests/utils_test.bal:253]: Can we add some tests to make sure that the equality check actually happens in constant time?

File: ballerina/Dependencies.toml [hunk 3]
```diff
@@ -107,7 +107,7 @@ modules = [
 [[package]]
 org = "ballerina"
 name = "time"
-version = "2.7.0"
+version = "2.8.0"
```
Comment (gayaldassanayake) [inline ballerina/Dependencies.toml]: Can you check into why the `time` version has been bumped here? Let's try to keep it in the original 2.7. If this was built with ballerina dist - 2201.12.0, time version shouldn't bump a minor

File: native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Utils.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * @since 2.11.0
+ */
+public class Compare {
+
+    private Compare() {}
+
+    public static boolean equalByteConstantTime(BArray value, BArray expectedValue) {
+        byte[] valueBytes = value.getBytes();
+        byte[] expectedBytes = expectedValue.getBytes();
+        return MessageDigest.isEqual(valueBytes, expectedBytes);
```
Comment (gayaldassanayake) [inline native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Utils.java:45]: Is MessageDigest.isEqual constant time?

File: native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Utils.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * Functions for constant-time comparison operations.
+ *
+ * @since 2.11.0
+ */
+public class Compare {
+
+    private Compare() {}
+
+    public static boolean equalByteConstantTime(BArray value, BArray expectedValue) {
```
Comment (gayaldassanayake) [inline native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Utils.java:42]: Add a doc comment to to public constructs

File: native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Utils.java (hunk 4 above)
Comment (NadhiyaNashath) [reply native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Utils.java:45]: Yes..
https://docs.oracle.com/en/java/javase/23/docs//api/java.base/java/security/MessageDigest.html#isEqual(byte%5B%5D,byte%5B%5D):~:text=the%20calculation%20time%20depends%20only%20on%20the%20length%20of%20digesta.%20It%20does%20not%20depend%20on%20the%20length%20of%20digestb%20or%20the%20contents%20of%20digesta%20and%20digestb.

File: ballerina/Dependencies.toml (hunk 3 above)
Comment (NadhiyaNashath) [reply ballerina/Dependencies.toml]: Resolved: https://github.com/ballerina-platform/module-ballerina-crypto/pull/624/changes/14ca840cead8de392dd6288464b3dd75c2f654eb

File: ballerina/tests/compare_test.bal (hunk 1 above)
Comment (NadhiyaNashath) [reply ballerina/tests/compare_test.bal]: Fixed: https://github.com/ballerina-platform/module-ballerina-crypto/pull/624/changes/b6b3511a0c28a3792af1f651d58b81699ff88733

File: ballerina/tests/compare_test.bal [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+@test:Config {}
+isolated function testEqualConstantTimeWithHmacBytes() {
+    byte[] input = "Hello Ballerina".toBytes();
+    byte[] key = "secret".toBytes();
+    byte[]|error value = hmacSha256(input, key);
+    byte[]|error expectedValue = hmacSha256(input, key);
+    if value is byte[] && expectedValue is byte[] {
+        test:assertTrue(equalConstantTime(value, expectedValue));
+    }
```
[the remark this answers is not in the document]
Comment (NadhiyaNashath) [reply ballerina/tests/compare_test.bal]: https://github.com/ballerina-platform/module-ballerina-crypto/pull/624/changes/b6b3511a0c28a3792af1f651d58b81699ff88733

File: ballerina/tests/utils_test.bal (hunk 2 above)
Comment (NadhiyaNashath) [reply ballerina/tests/utils_test.bal:253]: Added: https://github.com/ballerina-platform/module-ballerina-crypto/pull/624/changes/b6b3511a0c28a3792af1f651d58b81699ff88733

File: native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Utils.java (hunk 5 above)
Comment (NadhiyaNashath) [reply native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Utils.java:42]: Added: https://github.com/ballerina-platform/module-ballerina-crypto/pull/624/changes/b6b3511a0c28a3792af1f651d58b81699ff88733

File: ballerina/tests/utils_test.bal [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    byte[] b = "warmup-alpha".toBytes();
+    byte[] c = "warmup-XXXXX".toBytes();
+    foreach int i in 0 ..< TIMING_WARMUP {
+        _ = equalConstantTime(a, b);
+        _ = equalConstantTime(a, c);
+    }
+    // Warm up the large-array code path so the JIT compiles it before measurements begin.
+    byte[] large = makeLargeArray(100000);
+    byte[] largeDiff = large.clone();
+    largeDiff[0] = largeDiff[0] == 0 ? 1 : 0;
```
Comment (gayaldassanayake) [inline ballerina/tests/utils_test.bal:54]: Why are we doing this? - largeDiff[0] = largeDiff[0] == 0 ? 1 : 0;

Comment (NadhiyaNashath) [reply ballerina/tests/utils_test.bal:54]: It's to flip the byte value to a different one so that we can simulate a scenario where the first byte is different

File: ballerina/tests/compare_test.bal [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+    decimal diff = t1 > t2 ? t1 - t2 : t2 - t1;
+    decimal larger = t1 > t2 ? t1 : t2;
+    decimal ratio = larger == 0d ? 0d : diff / larger;
+    decimal diffMs = diff * 1000d;
+    return ratio > TIMING_RATIO_THRESHOLD && diffMs > TIMING_ABS_THRESHOLD_MS;
+}
+
+@test:Config {}
+function testEqualConstantTimeTimingDifferAtStartVsEnd() {
+    timingWarmup();
```
Comment (gayaldassanayake) [inline ballerina/tests/compare_test.bal]: We should do this using `@beforeeach` or `@beforegroup` - check and use the correct annotation.

File: ballerina/tests/compare_test.bal [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+        _ = equalConstantTime(base, altB);
+        totalB += time:monotonicNow() - t2;
+    }
+    return [totalA, totalB];
+}
+
+function isSuspiciousTiming(decimal t1, decimal t2) returns boolean {
+    decimal diff = t1 > t2 ? t1 - t2 : t2 - t1;
+    decimal larger = t1 > t2 ? t1 : t2;
+    decimal ratio = larger == 0d ? 0d : diff / larger;
```
Comment (gayaldassanayake) [inline ballerina/tests/compare_test.bal]: We should assert that both values are not zero, hence larger is not zero.

File: ballerina/tests/compare_test.bal [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+
+    test:assertFalse(
+        isSuspiciousTiming(tStart, tEnd),
+        "Timing gap detected: differ-at-start took " + (tStart * 1000d).toString()
+            + "ms vs differ-at-end " + (tEnd * 1000d).toString() + "ms over "
+            + TIMING_ITERATIONS.toString() + " iterations"
+    );
+}
+
+// Verifies match and mismatch at the midpoint take the same time (no early exit on equality).
```
Comment (gayaldassanayake) [inline ballerina/tests/compare_test.bal]: Add this comment within the @test:Config annotation as a description

File: changelog.md [hunk 11]
```diff
@@ -5,6 +5,11 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 
 ## [Unreleased]
 
+## [2.11.0] - 2026-04-03
```
Comment (gayaldassanayake) [inline changelog.md]:
```suggestion
## [Unreleased]
```

File: ballerina/tests/compare_test.bal (hunk 9 above)
Comment (NadhiyaNashath) [reply ballerina/tests/compare_test.bal]: Addressed: https://github.com/ballerina-platform/module-ballerina-crypto/pull/624/changes/08f9ae57ddff137756aeab0da8ba63bdb2727aa6

File: ballerina/tests/compare_test.bal (hunk 8 above)
Comment (NadhiyaNashath) [reply ballerina/tests/compare_test.bal]: Addressed: https://github.com/ballerina-platform/module-ballerina-crypto/pull/624/changes/08f9ae57ddff137756aeab0da8ba63bdb2727aa6

File: ballerina/tests/compare_test.bal (hunk 10 above)
Comment (NadhiyaNashath) [reply ballerina/tests/compare_test.bal]: Addressed: https://github.com/ballerina-platform/module-ballerina-crypto/pull/624/changes/08f9ae57ddff137756aeab0da8ba63bdb2727aa6

**Review by gayaldassanayake — DISMISSED** [review]

File: native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Compare.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+import io.ballerina.runtime.api.values.BArray;
+
+import java.security.MessageDigest;
+
+/**
+ * Functions for constant-time comparison operations.
+ *
+ * @since 2.11.0
+ */
+public class Compare {
```
Comment (TharmiganK) [inline native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Compare.java]: The class can be final?
Can't we come up with the better name for the class other than `Compare`?

**Comment by TharmiganK** [comment]
Can we put the time comparison between equals and this new API for a test sample (a 1KB hash value where the mismatch happens in the middle)?

File: native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Compare.java (hunk 12 above)
Comment (NadhiyaNashath) [reply native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Compare.java]: Would something like  ConstantTime or something with the suffix Utils be suitable?

Comment (TharmiganK) [reply native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Compare.java]: Utils might be suitable. It allows us to add any other utility functions required for cryptographic operations

Comment (NadhiyaNashath) [reply native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Compare.java]: CryptoUtils ? or just plain Utils would suffice?

Comment (TharmiganK) [reply native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Compare.java]: Utils

Comment (NadhiyaNashath) [reply native/src/main/java/io/ballerina/stdlib/crypto/nativeimpl/Compare.java]: Resolved in this commit: https://github.com/ballerina-platform/module-ballerina-crypto/pull/624/changes/ea5dbb40988200f84a0fd0f3912c27e73171d470

**Comment by TharmiganK** [comment]
Lets update the specification as well: https://github.com/ballerina-platform/module-ballerina-crypto/blob/master/docs/spec/spec.md

**Review by TharmiganK — APPROVED** [review]
LGTM
